src/devices/src/virtio/net/simple_handler.rs

```diff
@@ -0,0 +1,188 @@
+    // because many situations are not really recoverable. We should consider reporting them based
+    // on the  metrics/events solution when they appear, and not propagate them further unless
+    // it's really useful/necessary.
+    fn write_frame_to_guest(&mut self) -> result::Result<bool, Error> {
+        let num_bytes = self.rxbuf_current;
+
+        let mut chain = match self.rxq.iter()?.next() {
+            Some(c) => c,
+            _ => return Ok(false),
+        };
+
+        let mut count = 0;
+        let buf = &mut self.rxbuf[..num_bytes];
+
+        while let Some(desc) = chain.next() {
+            let left = buf.len() - count;
+
+            if left == 0 {
+                break;
+            }
+
+            let len = cmp::min(left, desc.len() as usize);
+            chain
+                .memory()
+                .write_slice(&buf[count..count + len], desc.addr())
+                .map_err(Error::GuestMemory)?;
+
+            count += len;
+        }
+
+        if count != buf.len() {
+            // The frame was too large for the chain.
+            warn!("rx frame too large");
+        }
+
+        self.rxq.add_used(chain.head_index(), count as u32)?;
+
+        self.rxbuf_current = 0;
+
+        Ok(true)
+    }
+
+    pub fn process_tap(&mut self) -> result::Result<(), Error> {
+        loop {
+            if self.rxbuf_current == 0 {
+                match self.tap.read(&mut self.rxbuf) {
+                    Ok(n) => self.rxbuf_current = n,
+                    Err(_) => {
+                        // TODO: Do something (logs, metrics, etc.) in response to an error when
+                        // reading from tap. EAGAIN means there's nothing available to read anymore
+                        // (because we open the TAP as non-blocking).
+                        break;
+                    }
+                }
+            }
+
+            if !self.write_frame_to_guest()? {
+                if !self.rxq.enable_notification()? {
+                    break;
+                }
+            }
+        }
+
+        if self.rxq.needs_notification()? {
+            self.driver_notify.signal_used_queue(RXQ_INDEX);
+        }
+
+        Ok(())
+    }
+
+    fn send_frame_from_chain(
+        &mut self,
+        chain: &mut DescriptorChain<M>,
+    ) -> result::Result<u32, Error> {
+        let mut count = 0;
+
+        while let Some(desc) = chain.next() {
+            let left = self.txbuf.len() - count;
+            let len = desc.len() as usize;
+
+            if len > left {
+                warn!("tx frame too large");
+                break;
+            }
+
+            chain
+                .memory()
+                .read_slice(&mut self.txbuf[count..count + len], desc.addr())
+                .map_err(Error::GuestMemory)?;
+
+            count += len;
+        }
+
+        self.tap.write(&self.txbuf[..count]).map_err(Error::Tap)?;
```

Comment:
How should we treat the case when less than count bytes could've been written to the tap device? I guess we can add at least a (TODO) comment about this :-?.